### PR TITLE
[DIR-1644] UI support for typescript-workflows

### PR DESCRIPTION
**Pull request by sebxian**
## Description

* [x] Finish linear todos (run button, flowchart)
* [x] Solve problem when switching editor instances
* [ ] Update `ui/src/assets/ts/workflow.d.ts` with the correct types
* [ ] Tests in CI need to be fixed in branch `typescript` (resulting from backend changes)
* [ ] UI e2e tests should be added

### Scope of this PR
* The LayoutSwitcher is disabled for ts-workflows. There is a follow-up issue to implement flowcharts for ts-workflows (DIR-1646). I decided to ignore the stored editor layout setting rather than changing it. So the editor will still remember the layout of the last yaml workflow when you open a yaml-workflow.
* No support yet for input forms under "run workflow" for ts-workflows (to be solved in follow up DIR-1668).

### Implementation notes
* I decided to place the workflow.d.ts in assets (similar to the translation files). Practically, it makes no big difference, but I found it more logical to keep this file somewhat separate from our ui code, because it is not relevant here. According to Jens, it will remain fairly small (about 15 functions overall that will be defined).
* It wasn't easy to find out how exactly to use the Monaco Editor API to achieve our goals. See [these notes](https://gist.github.com/sebxian/59bcd2dcd7e88a8c041efc7a8e62c14a) for more information about the implementation I ended up chosing.
* I tried to keep the Editor design component as context-neutral as possible. So it is now possible to pass in "tsLibs" (our typescript definitions) and the decision of what libs to pass down is made in the context-specific parent components (in this case, the WorkflowEditor).
* All editor instances use the same language server. The extra type definitions are globally applied to the "typescript" language, not per document. Hence, it was complicated to find out how to best set and clear the tsLibs. In the end, I chose the `useEffect` method after trying other methods. I will summarize the most important points here, but this is a topic worth discussing in person at a later time.
  * We need to clear the extraLibs when unmounting the workflow editor. Otherwise, the typescript definitions will still be known in new editor windows that use the typescript language. In theory, this could be a problem. In practice, at the current time, workflows are the only use case where we have typescript, so they would never be visible out of context.
  * I first tried to set the types whenever a new editor is instantiated, overwriting the previous config. But this would also happen when a new editor is opened in a modal in top of an existing editor, and as a result, the config would also be changed for the existing editor. The useEffect method avoids this problem. But if we ever have the use case where two editors that use typescript exist at the same time, the same typescript context will be active in both of them. 

### Notes for testing

To create a TS workflow, use this CURL (adjust host and namespace as needed):

```
curl --request POST \
  --url http://localhost:3000/api/v2/namespaces/test/files/ \
  --header 'Content-Type: application/json' \
  --header 'User-Agent: Insomnia/2023.5.3' \
  --data '{
	"name":"my-typescript.ts",
	"data":"ZnVuY3Rpb24gdmFsdWUoKSB7CgoJY29uc3QgZmlsZU9uZSA9IGdldEZpbGUoewoJCW5hbWU6ICIvbXlmaWxlLnR4dCIsCgkJcGVybWlzc2lvbjogNzU1LAoJCXNjb3BlOiAic2hhcmVkIiwKCX0pOwkgIAogICAgCiAgICB2YXIgcyA9IGdldFNlY3JldCh7IG5hbWU6ICJoZWxsby13b3JsZCJ9KQoKICAgIHZhciByID0gaHR0cFJlcXVlc3QoCiAgICAgICAgewogICAgICAgICAgICBtZXRob2Q6ICJQT1NUIiwKICAgICAgICAgICAgdXJsOiAiaHR0cDovLzEyNy4wLjAuMTolZCIKICAgICAgICB9CiAgICApCgogICAgdmFyIGZuID0gc2V0dXBGdW5jdGlvbih7CgkJaW1hZ2U6ICJsb2NhbGhvc3Q6NTAwMC9oZWxsbyIKCX0pCgogICAgdmFyIGZuID0gc2V0dXBGdW5jdGlvbjIoewoJCWltYWdlOiAibG9jYWxob3N0OjUwMDAvaGVsbG8iCgl9KQoKfQ==",
	"type":"workflow",
	"mimeType":"application/x-typescript"
}'
```

When editing a typescript-workflow, the editor should recognize `getFiles()` like this:

image

When navigating to route after this and editing a JavaScript plugin, the `getFiles()` method should not be known:

image

This will confirm the typescript info is no longer present when opening a non-workflow file.

## Checklist

- [ ] Documentation updated if required
- [ ] Test coverage is appropriate
      
## Checklist Internal

- [x] Linear issue linked (e.g. [DIR-XXXX] pull request title)
- [x] Has the PR been labeled

### CODE REVIEW

**Review by stefan-kracht — CHANGES_REQUESTED** [review]
This looks very good! I made some tiny suggestions in the code. 

In the new workflow modal, we can now end up in a state where the filename placeholder "suggests" a filename that does not comply with the workflow type. Like in the screenshot, where typescript is selected and the placeholder shows `workflow-name.yaml`. One easy way out would be to only show worklow-name.


CleanShot 2024-07-22 at 10 54 40@2x

**Comment by sebxian** [comment]
> In the new workflow modal, we can now end up in a state where the filename placeholder "suggests" a filename that does not comply with the workflow type. Like in the screenshot, where typescript is selected and the placeholder shows `workflow-name.yaml`. One easy way out would be to only show worklow-name.

Good point! Fixed in c97273c93, it now generates a placeholder matching the type. Some extra logic required but I think it's good to show users the naming pattern in the placeholder instead of hiding the extension.

File: ui/src/pages/namespace/Explorer/Tree/utils.ts [hunk 1]
```diff
@@ -10,3 +10,35 @@ export const addYamlFileExtension = (name: string) => {
...
+};
+
+/**
+ * Will add the provided extension to the name. Will strip (partial) matches
+ * of the extension first to avoid duplicating them.
+ * @param name
+ * @param extension
+ * @returns name with extension
+ */
+export const addFileExtension = (name: string, extension: string) => {
```
[the remark this answers is not in the document]
Comment (sebxian) [reply ui/src/pages/namespace/Explorer/Tree/utils.ts]: done in d172be808

File: ui/src/pages/namespace/Explorer/Tree/__tests__/utils.test.ts [hunk 2]
```diff
@@ -19,11 +22,58 @@ describe("addYamlFileExtension", () => {
...
+
+describe("stripFileExtension", () => {
+  const subject = stripFileExtension;
+  test("it strips the provided extension from the name", () => {
+    expect(subject("name.foo.ts", "foo.ts")).toBe("name");
+  });
+  test("it strips the last segment from the file name if it matches a part of the provided extension", () => {
+    expect(subject("name.ts", "foo.ts")).toBe("name");
+    expect(subject("name.foo", "foo.ts")).toBe("name");
+  });
```
[the remark this answers is not in the document]
Comment (sebxian) [reply ui/src/pages/namespace/Explorer/Tree/__tests__/utils.test.ts:42]: I added this test case, is this what you mean?

```
  test("it does nothing if the provided extension isn't contained in the file name", () => {
    expect(subject("name.ts", "notfound")).toBe("name.ts");
  });
```

To be honest I am not sure I understand the point though.

[the remark this answers is not in the document]
Comment (stefan-kracht) [reply ui/src/pages/namespace/Explorer/Tree/__tests__/utils.test.ts:42]: Yes, that's what I meant. For unit tests, I like to cover as many of the possible outcomes of a function as possible, as it becomes very handy when refactoring. One possible outcome that was not covered is that this function will not change the input string at all.

**Review by stefan-kracht — APPROVED** [review]

**Pull request merged**